feat: add auto share document on assignment

Thunderbottom opened the pull request:
Add a standard field in all DocTypes to allow automatic document sharing whenever a user is assigned a document. Whenever the user is unassigned, the document will be unshared with the user.

**Share Setting in DocType:**

![image](https://user-images.githubusercontent.com/11243138/88381980-2880c300-cdc5-11ea-8827-c31917877f4b.png)

Whenever the field above is enabled, all the documents for that DocType will automatically be shared with the User that the document is assigned to. The share permissions can be set, and is applicable for all documents.

Thunderbottom committed:
feat: add auto share document on assignment

add a standard field in all doctypes to allow automatic document sharing
whenever a user is assigned a document. whenever the user is unassigned,
the document is unshared with the user.

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

Thunderbottom committed:
fix: only validate auto share on assignment if enabled

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

kennethsequeira commented:
@Thunderbottom can you add an option in System Settings as well so a system-wide behaviour may be implemented.

prssanna commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -517,12 +524,51 @@
+  {
+   "default": "1",
+   "depends_on": "auto_share_on_assignment",
+   "fieldname": "allow_read",
```
```suggestion
   "fieldname": "allow_read_on_share",
```

prssanna commented on frappe/desk/doctype/todo/todo.py:
```diff
@@ -33,6 +34,16 @@ def validate(self):
+						read=doc_meta.allow_read, write=doc_meta.allow_write,
+						share=doc_meta.allow_share)
+			else:
+				remove(self.reference_type, self.reference_name, self.owner)
```
There's code that does this in `assign_to.py`. This can also be moved there.
If `auto_share_on_assignment` is disabled it still shares the document in `assign_to.py`

prssanna commented on frappe/core/doctype/doctype/doctype.py:
```diff
@@ -643,6 +644,15 @@ def validate_name(self, name=None):
 
+	def validate_auto_share_on_assignment(self):
+		if self.auto_share_on_assignment:
+			if not (self.issingle or self.istable or self.is_tree):
```
why `is_tree`?

Thunderbottom replied:
dont know, should we allow sharing and assignment on tree doctypes?

prssanna commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -517,12 +524,51 @@
+   "depends_on": "eval:!doc.istable && !doc.issingle",
+   "fieldname": "sharing_rules_section",
+   "fieldtype": "Section Break",
+   "label": "Sharing Rules"
```
Sharing Rules gives the impression that these properties will be applied whenever a document is shared. Maybe we could label it Assignment Permission Rules.

prssanna commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -517,12 +524,51 @@
+  {
+   "default": "0",
+   "depends_on": "auto_share_on_assignment",
+   "fieldname": "allow_write",
```
```suggestion
   "fieldname": "allow_write_for_assigned_users",
```
The field name gives the impression that all users will have write permission for the document

prssanna commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -517,12 +524,51 @@
+   "label": "Sharing Rules"
+  },
+  {
+   "default": "0",
```
Default should be 1. Also not sure if there is a use case for disabling it. Maybe we only need the fields `allow_write_for_assigned_users` and `allow_share_for_assigned_users`.

Thunderbottom committed:
feat: add role-based sharing on assignment

sharing permissions can be specified based on permissions, instead of
granting everyone all rights on share.

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

Thunderbottom committed:
chore: fix sider issue with semicolon

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

Thunderbottom committed:
fix: add table structure for share permissions

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

Thunderbottom committed:
chore: remove redundant variable assignments

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

Thunderbottom committed:
fix: use get on meta attribute

fixes issues on migrate:

AttributeError: 'Meta' object has no attribute 'auto_share_on_assignment'

Signed-off-by: Chinmay D. Pai <chinmaydpai@gmail.com>

deepeshgarg007 committed:
fix: reload doctype and share_permissions

rmehta commented:
This is a very complex implementation. This should be the default behaviour (I have a vague memory if implementing this myself)

prssanna commented:
@rmehta, yeah I think a setting for giving write and maybe share permission to the assigned user is enough (in addition to read which is the default behaviour). I don't think role based permission is required

Thunderbottom commented:
I believe that role permissions are required here since documents need to be shared based on roles. A simple check would do to assign write/share permissions with the user, but not all documents are supposed to be shared with everyone with the same permissions.

Previously if the user did not have access to the document, the document would be shared with a read permission with the user. That only works when done through the frontend. This PR handles auto-assignments as well, sharing documents with appropriate permissions. Yeah, it is a complex implementation, but so is our role permissions system.

rmehta commented:
@Thunderbottom I agree with @prssanna. At some point the cost of configuration becomes higher than the value of using the feature. This feels like one such case. Let us implement the best default behaviour - once you have assignment, you should be given a read+write share - seems a reasonable assumption. Also we can move this into the ORM if it is not implemented correctly.

prssanna commented:
@Thunderbottom any update on this? Can we remove role based permissions?
